crates/sui-verifier/src/char_type_verifier.rs

```diff
@@ -0,0 +1,227 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! A module can define a "characteristic type", that is a type that is instantiated only once, and
+//! this property is enforced by the system. We define a characteristic type as a struct type that
+//! has the same name as the module that defines it but with all the letter capitalized, and
+//! possessing certain special properties specified below (please note that by convention, "regular"
+//! struct type names are expressed in camel case).  In other words, if a module defines a struct
+//! type whose name is the same as the module name, this type MUST possess these special properties,
+//! otherwise the module definition will be considered invalid and will be rejected by the
+//! validator:
+//!
+//! - it has only one ability: drop
+//! - it has only one arbitrarily named field of type boolean (since Move structs cannot be empty)
+//! - its definition does not involve type parameters
+//! - its only instance in existence is passed as an argument to the module initializer
+//! - it is never instantiated anywhere in its defining module
+
+use move_binary_format::{
+    access::ModuleAccess,
+    binary_views::BinaryIndexedView,
+    file_format::{
+        Ability, AbilitySet, Bytecode, CompiledModule, FunctionDefinition, FunctionHandle,
+        SignatureToken, StructDefinition, StructHandle, TypeSignature,
+    },
+};
+use move_core_types::{ident_str, language_storage::ModuleId};
+use sui_types::{
+    base_types::{TX_CONTEXT_MODULE_NAME, TX_CONTEXT_STRUCT_NAME},
+    error::ExecutionError,
+    SUI_FRAMEWORK_ADDRESS,
+};
+
+use crate::{verification_failure, INIT_FN_NAME};
+
+pub fn verify_module(module: &CompiledModule) -> Result<(), ExecutionError> {
+    // In Sui's framework code there is an exception to the characteristic type rule - we have a SUI
+    // type in the sui module but it is instantiated and the module has no initializer (the reason
+    // for it is that the SUI coin is only instantiated during genesis). It is easiest to simply
+    // special-case this module particularly that this is framework code and thus deemed correct.
+    if ModuleId::new(SUI_FRAMEWORK_ADDRESS, ident_str!("sui").to_owned()) == module.self_id() {
+        return Ok(());
+    }
+
+    let view = BinaryIndexedView::Module(module);
+    let mod_handle = view.module_handle_at(module.self_module_handle_idx);
+    let mod_name = view.identifier_at(mod_handle.name).as_str();
+    let struct_defs = &module.struct_defs;
+    let mut char_type_candidate = None;
+    for def in struct_defs {
+        let struct_handle = module.struct_handle_at(def.struct_handle);
+        let struct_name = view.identifier_at(struct_handle.name).as_str();
+        if struct_name.to_ascii_uppercase() == struct_name
+            && mod_name.to_ascii_uppercase() == struct_name
```

Comment:
Do you need both of these here? Isn't it just `mod_name.to_ascii_uppercase() == struct_name`?

Reply:
When I was attempting to write an explanation, I realized that you are right :-)